[Concurrency] isIsolatingCurrentContext - Improved Custom SerialExecutor isolation checking (#2716)
swiftlang/swift-evolution: proposals/NNNN-SerialExecutor-isIsolated.md

# Improved Custom SerialExecutor isolation checking for Concurrency Runtime 

* Proposal: [SE-NNNN](...)
* Author: [Konrad 'ktoso' Malawski](https://github.com/ktoso)
* Review Manager:
* Status:  **WIP**
* Review: TODO

## Introduction

> In [SE-0424: Custom isolation checking for SerialExecutor](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0424-custom-isolation-checking-for-serialexecutor.md) we introduced a way for custom executors implementing the `SerialExecutor` protocol to assert and and assume the static isolation if the dynamic check succeeded. This proposal extends these capabilities, allowing custom executors to not only "check and crash if assumption was wrong", but also check and act on the result of the check.

Reviewer: ```suggestion
In [SE-0424: Custom isolation checking for SerialExecutor](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0424-custom-isolation-checking-for-serialexecutor.md) we introduced a way for custom executors implementing the `SerialExecutor` protocol to assert and assume the static isolation if the dynamic check succeeded. This proposal extends these capabilities, allowing custom executors to not only "check and crash if assumption was wrong", but also check and act on the result of the check.
```

Currently all APIs available are using the `checkIsolated()` API which must crash if called from a context not managed by the serial executor it is invoked on. This method is often implemented using `dispatchPrecondition()` when the serial executor is backed using `Dispatch` or custom `fatalError()` messages otherwise:

```swift
final class ExampleExecutor: SerialExecutor {
  func checkIsolated() {
    dispatchPrecondition(condition: .onQueue(self.queue))
  }
}
```

This approach is better than not being able to participate in the checks at all (i.e. this API allows for advanced thread/queue sharing between actor and non-actor code), but it has two severe limitations:

- the crash **messages** offered by these `checkIsolated()`  crashes **are often sub-optimal** and confusing
  - messages often don't include crucial information about which actor/executor the calling context was _actually_ executing on. Offering only "expected [...]" messages, leading to hard to debug crashes.
- it is **impossible** for the Swift runtime to offer **isolation violation warnings**

>   - because the Swift runtime _must_ call into a custom executor to verify its isolation, the "pass or crash" method will crash, rather than inform the runtime that a violation ocured and we should warn about it.

Reviewer: ```suggestion
  - because the Swift runtime _must_ call into a custom executor to verify its isolation, the "pass or crash" method will crash, rather than inform the runtime that a violation occurred and we should warn about it.
```

Today, it is not possible for the Swift runtime to issue _warnings_ if something is detected to be on not the expected executor, but somehow we'd still like to continue without crashing the application. 

And for existing situations when `assumeIsolated()` is called and _should_ crash, by using this new API the Swift runtime will be able to provide _more informative_ messages, including all available context about the execution environment.

## Proposed solution

We propose to introduce a new `isIsolatingCurrentContext` protocol requirement to the `SerialExecutor` protocol:

```swift
protocol SerialExecutor {

  /// May be called by the Swift runtime before `checkIsolated()`
  /// in order to check for isolation violations at runtime, 
  /// and potentially issue isolation warnings.
  ///
  /// [...]
  func isIsolatingCurrentContext() -> Bool

  // existing API, since SE-0424
  @available(SwiftStdlib 6.0, *)
  func checkIsolated() // must crash if run on a context not managed by this serial executor
  
  // ... 
}

extension SerialExecutor { 
  /// Default implementation for backwards compatibility.
  func isIsolatingCurrentContext() -> Bool { false }
}
```

> The Swift runtime is free to call the `isIsolated` function whenever it wants to verify if the current context is apropriately isolated by some serial executor. 

Reviewer: ```suggestion
The Swift runtime is free to call the `isIsolated` function whenever it wants to verify if the current context is appropriately isolated by some serial executor. 
```

> In most cases implementing this new API is preferable to implementing `checkIsolated()`, as the Swift runtime is able to offer more detailed error messages when when an isolation failure detected by a call to `isIsolatingCurrentContext()` is detected.

Reviewer: ```suggestion
In most cases implementing this new API is preferable to implementing `checkIsolated()`, as the Swift runtime is able to offer more detailed error messages when an isolation failure detected by a call to `isIsolatingCurrentContext()` is detected.
```

  - invoke `expected.checkIsolated` which will crash :x: or pass :white_check_mark: depending on its internal checking.
- if neither `checkIsolated` or `isIsolatingCurrentContext` are available, 
  - :x: crash with a best effort message.

This proposal specifically adds the "if `isIsolatingCurrentContext` is available" branch into the existing logic for confirming the isolation expectation.

If `isIsolatingCurrentContext` is available, effectively it replaces `checkIsolated` because it does offer a sub-par error message experience and is not able to offer a warning if Swift would be asked to check the isolation but not crash upon discovering a violation.

### Enabling `isIsolatingCurrentContext` checking mode

Similar as complex equality in `SerialExecutors` this feature must be opted into by flagging it when the `UnownedSerialExecutor` value is returned from a serial executor's `asUnownedSerialExecutor()`.

Previously this was done by signalling the feature in the `UnownedSerialExecutor` initializer like this:

```swift
// Existing API
public func asUnownedSerialExecutor() -> UnownedSerialExecutor {
  UnownedSerialExecutor(complexEquality: self)
}
```

> Which enables the following `isSameExclusiveExecutionContext` check, which can only be used when a "current" executor is present, and cannot be used when running code outside of a Swift concurren ytask (!):

Reviewer 2: ```suggestion
Which enables the following `isSameExclusiveExecutionContext` check, which can only be used when a "current" executor is present, and cannot be used when running code outside of a Swift concurrency task (!):
```

```swift
// Existing API
public func isSameExclusiveExecutionContext(other: NaiveQueueExecutor) -> Bool {
  other.secretIdentifier == self.secretIdentifier
}
```

> In order to enable the the runtime to call into the `isIsolatingCurrentContext` the `UnownedSerialExecutor` **must** be constructed as follows:

Reviewer 3: ```suggestion
In order to enable the runtime to call into the `isIsolatingCurrentContext` the `UnownedSerialExecutor` **must** be constructed as follows:
```

If `isIsolatingCurrentContext` is available, effectively it replaces `checkIsolated` because it does offer a sub-par error message experience and is not able to offer a warning if Swift would be asked to check the isolation but not crash upon discovering a violation.

### Detecting the `isIsolatingCurrentContext` checking mode

The `isIsolatingCurrentContext` method effectively replaces the `checkIsolated` method, because it can answer the same question _if it is implemented_.

Some runtimes may not be able to implement a the returning `isIsolatingCurrentContext`, and they are not expected to implement the new protocol requirement.

The general guidance about which method to implement is to implement `isIsolatingCurrentContext` whenever possible. This method can be used by the Swift runtime in "warning mode". When running a check in this mode, the `checkIsolated` method cannot and will not be used because it would cause an unexpected crash.

> The presence of a non-default implementation of the `isIsolatingCurrentContext` protocol requirement. In other words, if there is an implementation of the requirement available _other than_ the default one provided in the concurrency library, the runtime will attempt to use this method _over_ the `checkIsolated` API. This allows for a smooth migration to the new API, and enables the use of this method in if the runtime would like issue a check that cannot cause a crash.

Reviewer 4: The first sentence is missing... something

Author: Thank you, fixed!

> The presence of a non-default implementation of the `isIsolatingCurrentContext` protocol witness is detected by the compiler and the runtime can detect this information in order to determine if the new function should be used for these checks.
